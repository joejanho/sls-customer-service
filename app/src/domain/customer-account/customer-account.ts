import * as customerAccountCreatedEvent from "app/src/events/customer-account-created";
import * as customerAccountUpdatedEvent from "app/src/events/customer-account-updated";

import {
  CreateCustomerAccountProps,
  NewCustomerAccountProps,
} from "app/src/models/customer-account-types";

import { AggregateRoot } from "app/src/entity/aggregate-root";
import { CustomerAccountDto } from "app/src/dto/customer-account";
import { DomainEvent } from "app/src/entity/domain-event";
import { schema } from "app/src/schemas/customer-account.schema";

export class CustomerAccount extends AggregateRoot<CreateCustomerAccountProps> {
  private constructor({
    id,
    created,
    updated,
    ...props
  }: CreateCustomerAccountProps) {
    super(props, id, created, updated);
  }

  public retrieveDomainEvents(): DomainEvent[] {
    const existingEvents = [...this.domainEvents];

    // clear down the existing events on the queue
    this.clearDomainEvents();

    return existingEvents;
  }

  public static createAccount(props: NewCustomerAccountProps): CustomerAccount {
    const customerAccountProps: CreateCustomerAccountProps = {
      firstName: props.firstName,
      surname: props.surname,
    };

    const instance: CustomerAccount = new CustomerAccount(customerAccountProps);
    instance.validate(schema);

    instance.addDomainEvent({
      event: instance.toDto(),
      eventName: customerAccountCreatedEvent.eventName,
      source: customerAccountCreatedEvent.eventSource,
      eventSchema: customerAccountCreatedEvent.eventSchema,
      eventVersion: customerAccountCreatedEvent.eventVersion,
    });

    return instance;
  }

  public toDto(): CustomerAccountDto {
    return {
      id: this.id,
      created: this.created,
      updated: this.updated,
      firstName: this.props.firstName,
      surname: this.props.surname,
    };
  }

  public static toDomain(raw: CustomerAccountDto): CustomerAccount {
    const instance = new CustomerAccount({
      ...raw,
    });
    return instance;
  }
}
