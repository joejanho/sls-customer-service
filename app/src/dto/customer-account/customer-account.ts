export type CreateCustomerAccountDto = {
  id?: string;
  created?: string;
  updated?: string;
  firstName: string;
  surname: string;
};

export type CustomerAccountDto = {
  id: string;
  created: string;
  updated: string;
  firstName: string;
  surname: string;
};

export type NewCustomerAccountDto = {
  firstName: string;
  surname: string;
};
