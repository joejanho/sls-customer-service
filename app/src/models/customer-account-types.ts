export type CreateCustomerAccountProps = {
  id?: string;
  created?: string;
  updated?: string;
  firstName: string;
  surname: string;
};

export type CustomerAccountProps = {
  id: string;
  created: string;
  updated: string;
  firstName: string;
  surname: string;
};

export type NewCustomerAccountProps = {
  firstName: string;
  surname: string;
};
