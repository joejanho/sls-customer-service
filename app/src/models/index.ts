export * from "./customer-account-types";
