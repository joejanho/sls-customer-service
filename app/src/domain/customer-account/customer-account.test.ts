import { NewCustomerAccountProps } from "app/src/models/customer-account-types";

import { CustomerAccount } from "app/src/domain/customer-account";
import { CustomerAccountDto } from "app/src/dto/customer-account";

let customer: CustomerAccountDto = {
  created: "2022-01-01T00:00:00.000Z",
  firstName: "joe",
  id: "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  surname: "janho",
  updated: "2022-01-01T00:00:00.000Z",
};

describe("customer-account", () => {
  beforeAll(() => {
    jest.useFakeTimers().setSystemTime(new Date("2022-01-01"));
  });

  afterAll(() => {
    jest.clearAllTimers();
    jest.clearAllMocks();
  });

  describe("create account", () => {
    it("should fail creating a new account with an invalid customer", () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: "Â±", // invalid
        surname: "any",
      };

      // act / assert
      expect(() =>
        CustomerAccount.createAccount(newCustomerAccount)
      ).toThrowErrorMatchingInlineSnapshot(
        `"[{\\"instancePath\\":\\"/firstName\\",\\"schemaPath\\":\\"#/properties/firstName/pattern\\",\\"keyword\\":\\"pattern\\",\\"params\\":{\\"pattern\\":\\"^[a-zA-Z]+$\\"},\\"message\\":\\"must match pattern \\\\\\"^[a-zA-Z]+$\\\\\\"\\"}]"`
      );
    });

    it("should create the new account successfully with a valid customer", () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: "joe",
        surname: "janho",
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);

      // assert
      expect(newCustomer).toMatchSnapshot();
    });
  });

  describe("toDto", () => {
    it("should create the correct dto", () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: "joe",
        surname: "janho",
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);

      // assert
      expect(newCustomer.toDto()).toMatchSnapshot();
    });
  });

  describe("toDomain", () => {
    it("should create a domain object based on a dto", () => {
      // arrange
      const customerAccountProps: CustomerAccountDto = {
        id: "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        firstName: "joe",
        surname: "janho",
        created: "2022-01-01T00:00:00.000Z",
        updated: "2022-01-01T00:00:00.000Z",
      };

      // act
      const customerAccount: CustomerAccount =
        CustomerAccount.toDomain(customerAccountProps);

      // assert
      expect(customerAccount).toMatchSnapshot();
    });
  });

  describe("add domain event", () => {
    it("should return the domain events on the object", () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: "joe",
        surname: "janho",
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);

      // assert
      expect(newCustomer.retrieveDomainEvents()).toMatchSnapshot();
    });

    it("should clear all domain events once retrieved", () => {
      // arrange
      const newCustomerAccount: NewCustomerAccountProps = {
        firstName: "joe",
        surname: "janho",
      };

      // act
      const newCustomer = CustomerAccount.createAccount(newCustomerAccount);
      newCustomer.retrieveDomainEvents(); // call this once to flush the events

      // assert
      expect(newCustomer.retrieveDomainEvents()).toEqual([]);
    });
  });
});
