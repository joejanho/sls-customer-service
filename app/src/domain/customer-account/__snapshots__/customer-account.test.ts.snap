// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customer-account add domain event should return the domain events on the object 1`] = `
Array [
  Object {
    "event": Object {
      "created": "2022-01-01T00:00:00.000Z",
      "firstName": "joe",
      "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
      "surname": "janho",
      "updated": "2022-01-01T00:00:00.000Z",
    },
    "eventDateTime": "2022-01-01T00:00:00.000Z",
    "eventName": "CustomerAccountCreated",
    "eventVersion": "1",
    "source": "com.customer-account-bank",
  },
]
`;

exports[`customer-account create account should create the new account successfully with a valid customer 1`] = `
CustomerAccount {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [
    Object {
      "event": Object {
        "created": "2022-01-01T00:00:00.000Z",
        "firstName": "joe",
        "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
        "surname": "janho",
        "updated": "2022-01-01T00:00:00.000Z",
      },
      "eventDateTime": "2022-01-01T00:00:00.000Z",
      "eventName": "CustomerAccountCreated",
      "eventVersion": "1",
      "source": "com.customer-account-bank",
    },
  ],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "firstName": "joe",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "surname": "janho",
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;

exports[`customer-account toDomain should create a domain object based on a dto 1`] = `
CustomerAccount {
  "_created": "2022-01-01T00:00:00.000Z",
  "_domainEvents": Array [],
  "_id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "_updated": "2022-01-01T00:00:00.000Z",
  "props": Object {
    "created": "2022-01-01T00:00:00.000Z",
    "firstName": "joe",
    "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
    "surname": "janho",
    "updated": "2022-01-01T00:00:00.000Z",
  },
}
`;

exports[`customer-account toDto should create the correct dto 1`] = `
Object {
  "created": "2022-01-01T00:00:00.000Z",
  "firstName": "joe",
  "id": "f39e49ad-8f88-448f-8a15-41d560ad6d70",
  "surname": "janho",
  "updated": "2022-01-01T00:00:00.000Z",
}
`;
